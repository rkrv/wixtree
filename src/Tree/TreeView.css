.tree-branch {
    margin-left: 1rem;
}

.tree-branch::before {
    content: '- ';
}
