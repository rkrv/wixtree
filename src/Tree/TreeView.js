import React, { Component } from 'react';
import TreeNode from './TreeNode';
import './TreeView.css';

export const MSG_NO_DATA = 'No tree data is loaded.';

export default class TreeView extends Component {
  static defaultProps = {
    data: null
  };

  renderRecursively() {
    if ( ! this.props.data.label) return null;
    return this.renderNodeRecursively(this.props.data);
  }

  renderNodeRecursively(node, depth = 0, key = 0) {
    const out = [];
    let counter = key;

    out.push(<TreeNode key={ counter++ } depth={ depth } data={ node } />);
    out.push(...node.children.map(child => this.renderNodeRecursively(child, depth + 1, counter++)));

    return out;
  }

  renderIteratively() {
    if ( ! this.props.data.label) return null;

    const queue = [{ depth: 0, data: this.props.data }];
    const nodes = [];
    let counter = 0;

    while (queue.length) {
      const node = queue.shift();

      nodes.push(<TreeNode key={ counter } depth={ node.depth } data={ node.data } />);
      counter++;

      if (node.data.children.length) {
        const children = node.data.children.map(child => ({ depth: node.depth + 1, data: child }));
        queue.unshift(...children);
      }
    }

    return nodes;
  }

  render() {
    if ( ! this.props.data) {
      return MSG_NO_DATA;
    }

    return (
      <div className="split-panel">
        <div>
          { this.renderRecursively() }
        </div>

        <div>
          { this.renderIteratively() }
        </div>
      </div>
    );
  }
}
