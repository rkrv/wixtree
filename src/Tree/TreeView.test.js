import React from 'react';
import TreeView, { MSG_NO_DATA } from './TreeView';
import treeJson from './tree';

it('renders without crashing', () => {
  const wrapper = shallow(<TreeView />);
  expect(wrapper).toHaveLength(1);
});

it('has reasonable defaultProps', () => {
  expect(TreeView.defaultProps.data).toBeNull();
});

it('renders a message when no tree data exists', () => {
  const wrapper = shallow(<TreeView />);
  expect(wrapper.text()).toBe(MSG_NO_DATA);
});

describe('with data', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<TreeView data={ treeJson } />);
  });

  afterEach(() => {
    wrapper.unmount();
    wrapper = null;
  });

  it('renders recursively', () => {
    expect(wrapper.instance().renderRecursively()).toMatchSnapshot();
  });

  it('renders iteratively', () => {
    expect(wrapper.instance().renderIteratively()).toMatchSnapshot();
  });
});
