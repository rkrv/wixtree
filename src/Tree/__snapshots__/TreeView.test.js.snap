// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with data renders iteratively 1`] = `
Array [
  <TreeNode
    data={
        Object {
            "children": Array [
              Object {
                "children": Array [],
                "label": "Demi Meower",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "label": "Chairman Meow",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "label": "Purrfect",
                      },
                      Object {
                        "children": Array [],
                        "label": "Thunder Paws",
                      },
                    ],
                    "label": "Cat-titude",
                  },
                ],
                "label": "Fleas Witherspoon",
              },
              Object {
                "children": Array [],
                "label": "Puma Thurman",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "label": "Katy Purry",
                      },
                    ],
                    "label": "Miraclaw",
                  },
                  Object {
                    "children": Array [],
                    "label": "Anderson Pooper",
                  },
                ],
                "label": "Whispurr",
              },
            ],
            "label": "Root node",
          }
    }
    depth={0}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [],
            "label": "Demi Meower",
          }
    }
    depth={1}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [
              Object {
                "children": Array [],
                "label": "Chairman Meow",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "label": "Purrfect",
                  },
                  Object {
                    "children": Array [],
                    "label": "Thunder Paws",
                  },
                ],
                "label": "Cat-titude",
              },
            ],
            "label": "Fleas Witherspoon",
          }
    }
    depth={1}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [],
            "label": "Chairman Meow",
          }
    }
    depth={2}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [
              Object {
                "children": Array [],
                "label": "Purrfect",
              },
              Object {
                "children": Array [],
                "label": "Thunder Paws",
              },
            ],
            "label": "Cat-titude",
          }
    }
    depth={2}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [],
            "label": "Purrfect",
          }
    }
    depth={3}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [],
            "label": "Thunder Paws",
          }
    }
    depth={3}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [],
            "label": "Puma Thurman",
          }
    }
    depth={1}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "label": "Katy Purry",
                  },
                ],
                "label": "Miraclaw",
              },
              Object {
                "children": Array [],
                "label": "Anderson Pooper",
              },
            ],
            "label": "Whispurr",
          }
    }
    depth={1}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [
              Object {
                "children": Array [],
                "label": "Katy Purry",
              },
            ],
            "label": "Miraclaw",
          }
    }
    depth={2}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [],
            "label": "Katy Purry",
          }
    }
    depth={3}
/>,
  <TreeNode
    data={
        Object {
            "children": Array [],
            "label": "Anderson Pooper",
          }
    }
    depth={2}
/>,
]
`;

exports[`with data renders recursively 1`] = `
Array [
  <TreeNode
    data={
        Object {
            "children": Array [
              Object {
                "children": Array [],
                "label": "Demi Meower",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "label": "Chairman Meow",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "label": "Purrfect",
                      },
                      Object {
                        "children": Array [],
                        "label": "Thunder Paws",
                      },
                    ],
                    "label": "Cat-titude",
                  },
                ],
                "label": "Fleas Witherspoon",
              },
              Object {
                "children": Array [],
                "label": "Puma Thurman",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "label": "Katy Purry",
                      },
                    ],
                    "label": "Miraclaw",
                  },
                  Object {
                    "children": Array [],
                    "label": "Anderson Pooper",
                  },
                ],
                "label": "Whispurr",
              },
            ],
            "label": "Root node",
          }
    }
    depth={0}
/>,
  Array [
    <TreeNode
      data={
            Object {
                  "children": Array [],
                  "label": "Demi Meower",
                }
      }
      depth={1}
/>,
  ],
  Array [
    <TreeNode
      data={
            Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "label": "Chairman Meow",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [],
                          "label": "Purrfect",
                        },
                        Object {
                          "children": Array [],
                          "label": "Thunder Paws",
                        },
                      ],
                      "label": "Cat-titude",
                    },
                  ],
                  "label": "Fleas Witherspoon",
                }
      }
      depth={1}
/>,
    Array [
      <TreeNode
        data={
                Object {
                        "children": Array [],
                        "label": "Chairman Meow",
                      }
        }
        depth={2}
/>,
    ],
    Array [
      <TreeNode
        data={
                Object {
                        "children": Array [
                          Object {
                            "children": Array [],
                            "label": "Purrfect",
                          },
                          Object {
                            "children": Array [],
                            "label": "Thunder Paws",
                          },
                        ],
                        "label": "Cat-titude",
                      }
        }
        depth={2}
/>,
      Array [
        <TreeNode
          data={
                    Object {
                              "children": Array [],
                              "label": "Purrfect",
                            }
          }
          depth={3}
/>,
      ],
      Array [
        <TreeNode
          data={
                    Object {
                              "children": Array [],
                              "label": "Thunder Paws",
                            }
          }
          depth={3}
/>,
      ],
    ],
  ],
  Array [
    <TreeNode
      data={
            Object {
                  "children": Array [],
                  "label": "Puma Thurman",
                }
      }
      depth={1}
/>,
  ],
  Array [
    <TreeNode
      data={
            Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [],
                          "label": "Katy Purry",
                        },
                      ],
                      "label": "Miraclaw",
                    },
                    Object {
                      "children": Array [],
                      "label": "Anderson Pooper",
                    },
                  ],
                  "label": "Whispurr",
                }
      }
      depth={1}
/>,
    Array [
      <TreeNode
        data={
                Object {
                        "children": Array [
                          Object {
                            "children": Array [],
                            "label": "Katy Purry",
                          },
                        ],
                        "label": "Miraclaw",
                      }
        }
        depth={2}
/>,
      Array [
        <TreeNode
          data={
                    Object {
                              "children": Array [],
                              "label": "Katy Purry",
                            }
          }
          depth={3}
/>,
      ],
    ],
    Array [
      <TreeNode
        data={
                Object {
                        "children": Array [],
                        "label": "Anderson Pooper",
                      }
        }
        depth={2}
/>,
    ],
  ],
]
`;
